include/deal.II/fe/fe_enriched.h

```diff
@@ -0,0 +1,545 @@
+                       dealii::internal::FEValues::FiniteElementRelatedData<dim, spacedim> &output_data
+                      ) const;
+
+  virtual
+  void
+  fill_fe_face_values ( const typename Triangulation<dim, spacedim>::cell_iterator   &cell,
+                        const unsigned int face_no,
+                        const Quadrature<dim-1> &quadrature,
+                        const Mapping<dim, spacedim> &mapping,
+                        const typename Mapping<dim, spacedim>::InternalDataBase &mapping_internal,
+                        const dealii::internal::FEValues::MappingRelatedData<dim, spacedim> &mapping_data,
+                        const typename FiniteElement<dim,spacedim>::InternalDataBase &fe_internal,
+                        dealii::internal::FEValues::FiniteElementRelatedData<dim, spacedim> &output_data
+                      ) const;
+
+  virtual
+  void
+  fill_fe_subface_values (const typename Triangulation< dim, spacedim >::cell_iterator &cell,
+                          const unsigned int face_no,
+                          const unsigned int sub_no,
+                          const Quadrature<dim-1> &quadrature,
+                          const Mapping<dim, spacedim> &mapping,
+                          const typename Mapping< dim, spacedim >::InternalDataBase &mapping_internal,
+                          const dealii::internal::FEValues::MappingRelatedData<dim, spacedim> &mapping_data,
+                          const typename FiniteElement<dim,spacedim>::InternalDataBase &fe_internal,
+                          dealii::internal::FEValues::FiniteElementRelatedData<dim, spacedim> &output_data
+                         ) const;
+
+private:
+  /**
+   * This function sets up the index table for the system as well as @p
+   * restriction and @p prolongation matrices.
+   */
+  void initialize (const std::vector<const FiniteElement<dim,spacedim>*> &fes,
+                   const std::vector<unsigned int> &multiplicities);
+
+  /**
+   * The underlying FESystem object.
+   */
+  FESystem<dim,spacedim> fe_system;
+
+  /**
+   * After calling fill_fe_(face/subface_)values this function
+   * implements the chain rule to multiply stored shape values/gradient/hessians
+   * by those of enrichment function evaluated at quadrature points.
+   */
+  template <int dim_1>
+  void
+  multiply_by_enrichment (const Quadrature<dim_1> &quadrature,
+                          const typename FiniteElement<dim,spacedim>::InternalDataBase  &fe_internal,
+                          const internal::FEValues::MappingRelatedData<dim,spacedim>    &mapping_data,
+                          const typename Triangulation< dim, spacedim >::cell_iterator  &cell,
+                          internal::FEValues::FiniteElementRelatedData<dim,spacedim>    &output_data) const;
+
+  /**
+   * Thread mutex for protecting multiply_by_enrichment().
+   */
+  mutable Threads::Mutex mutex;
```

Comment:
I don't think I understand why you need this. The function is `const`, all of its input arguments are `const`, and there are no other `mutable` variables in the class. Consequently, there is nothing other than the mutable members of the `fe_internal` object that you could possibly modify, and that object is not accessed from different threads anyway. So I think you should be ok without the mutex.